### ISSUE TITLE
M2MResource pointers are NOT SAFE

### ISSUE THREAD

**Issue by Alex-EEE** [ARMmbed/simple-mbed-cloud-client#75]
All of the example repos use this design pattern:

/* Create resource */
    MbedCloudClientResource *variable;
    variable = client.create_resource("3201/0/5853", "variable");
    variable->set_value("assign new value");
    variable->methods(M2MMethod::GET | M2MMethod::PUT);

However, it's not documented that this is NOT SAFE AT ALL!!!  When a PUT is carried out on a resource, the mbed-client deletes the old resource and creates a new one.  For example, in mbed-client-lite, registry_remove_object() below invalidates and M2MResource pointers on the resource:

```
// PUT on a resource is a REGISTRY_OPERATION_REPLACE
curr_path.path_type = REGISTRY_PATH_RESOURCE;
            if (operation == REGISTRY_OPERATION_REPLACE) {
                /* remove everything under this object instance */
                registry_status_t status = registry_remove_object(registry, &curr_path, REGISTRY_REMOVE_FOR_REPLACEMENT);

                if (status != REGISTRY_STATUS_OK && status != REGISTRY_STATUS_NOT_FOUND) {
                    return REGISTRY_TLV_SERIALIZE_STATUS_GENERIC_ERROR;
                }
            }
            status = registry_deserialize_resource_tlv(registry, &curr_path, &til, operation);
```

Well, thank you very much guys for that little land mine.  And I get it, a PUT is essentially creating a new resource in the Lwm2m object, so it stands to reason you get a new resource in C++ as well.  But the example code implies that you can create resources and save off pointers to them and access them that way for the life of the M2MObject.

PLEASE DOCUMENT THIS....and set up your example code so that you encourage users to save off M2MObjectInstance pointers instead of M2MResource pointers.

**Comment by screamerbg** [comment]
@Alex-EEE Thanks for the investigation.

@janjongboom This one is for you.

**Comment by Alex-EEE** [comment]
In hindsight, I probably should have posted this on "mbed-cloud-client-example" since that's the more "mainline" code base and they do the exact same thing over there.  But since we got @janjongboom attention, maybe if he agrees with me he can post it over there and it will get more attention.

**Comment by janjongboom** [comment]
@Alex-EEE Thanks for reporting, this is a first for me. @yogpan01 can you confirm that this is the PUT behavior?

**Comment by yogpan01** [comment]
Hi @Alex-EEE,
I think you are mixing two completely different components. PDMC (mbed Cloud Client) doesnt have this behaviour and our simple-mbed-cloud-client are for that and not for Client Lite.
As for Client Lite, we will check the behaviour and fix if there is a bug on the handling.

**Comment by janjongboom** [comment]
@Alex-EEE yes, been digging through the code, and this seems unrelated to this repo. Not saying that this is not a valid issue with Client Lite, but I'll let @yogpan01 comment on that.

**Comment by yogpan01** [comment]
Hi @Alex-EEE,
We are now checking the PUT behaviour for Object, ObjectInstance, Resource and ResourceInstance on Client Lite.
We will update the findings and if there are defects like you posted, those will be fixed in upcoming releases.
